src/celeritas/optical/CerenkovDndxCalculator.hh

```diff
@@ -0,0 +1,152 @@
+#pragma once
+
+#include <cmath>
+
+#include "corecel/Assert.hh"
+#include "corecel/Macros.hh"
+#include "corecel/Types.hh"
+#include "celeritas/grid/GenericCalculator.hh"
+#include "celeritas/grid/GenericGridData.hh"
+#include "celeritas/grid/VectorUtils.hh"
+
+#include "CerenkovData.hh"
+#include "OpticalPropertyData.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Calculate the mean number of Cerenkov photons produced per unit length.
+ *
+ * The average number of photons produced is given by
+ * \f[
+   dN = \frac{\alpha z^2}{\hbar c}\sin^2\theta d\epsilon dx = \frac{\alpha
+ z^2}{\hbar c}\left(1 - \frac{1}{n^2\beta^2}\right) d\epsilon dx,
+ * \f]
+ * where \f$ n \f$ is the refractive index of the material, \f$ \epsilon \f$
+ * is the photon energy, and \f$ \theta \f$ is the angle of the emitted photons
+ * with respect to the incident particle direction, given by \f$ \cos\theta = 1
+ * / (\beta n) \f$. Note that in a dispersive medium, the index of refraction
+ * is an inreasing function of photon energy. The mean number of photons per
+ * unit length is given by
+ * \f[
+   dN/dx = \frac{\alpha z^2}{\hbar
+ c}\int_{\epsilon_\text{min}}^{\epsilon_\text{max}}d\epsilon\left(1 -
+ \frac{1}{n^2\beta^2}\right) = \frac{\alpha z^2}{\hbar
+ c}\left[\epsilon_\text{max} - \epsilon_\text{min} -
+ \frac{1}{\beta^2}\int_{\epsilon_\text{min}}^{\epsilon_\text{max}}
+ \frac{d\epsilon}{n^2(\epsilon)}\right].
+ * \f]
+ */
+class CerenkovDndxCalculator
+{
+  public:
+    // Construct from optical properties and Cerenkov angle integrals
+    inline CELER_FUNCTION
+    CerenkovDndxCalculator(OpticalPropertyRef const& properties,
+                           CerenkovRef const& shared,
+                           OpticalMaterialId material,
+                           real_type charge);
+
+    // Calculate the mean number of Cerenkov photons produced per unit length
+    inline CELER_FUNCTION real_type operator()(real_type inv_beta);
+
+  private:
+    OpticalPropertyRef const& properties_;
+    CerenkovRef const& shared_;
+    OpticalMaterialId material_;
+    real_type alpha_zsq_over_hbar_c_;
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from optical properties and Cerenkov angle integrals.
+ */
+CELER_FUNCTION
+CerenkovDndxCalculator::CerenkovDndxCalculator(
+    OpticalPropertyRef const& properties,
+    CerenkovRef const& shared,
+    OpticalMaterialId material,
+    real_type charge)
+    : properties_(properties)
+    , shared_(shared)
+    , material_(material)
+    , alpha_zsq_over_hbar_c_(1e14 * constants::alpha_fine_structure
+                             * constants::e_electron * ipow<2>(charge)
+                             / (constants::hbar_planck * constants::c_light))
```

Comment:
What is 1e14?? If charge is indeed an elementary unity then this is equivalent to:

```suggestion
    , alpha_zsq_over_hbar_c_(1e14 * constants::alpha_fine_structure
                             * ipow<2>(native_value_from(charge))
                             / (constants::hbar_planck * constants::c_light * constants::e_electron))
```
but I don't see where the charge shows up in the earlier equation either...

Reply:
Sorry yeah, one of the parts I still need to fix up :)